x/auth/middleware/recovery.go

```diff
@@ -54,19 +59,18 @@ func (txh recoveryTxHandler) DeliverTx(ctx context.Context, req tx.Request) (res
 		}
 	}()
 
-	// If BlockGasMeter() panics it will be caught by the above recover and will
-	// return an error - in any case BlockGasMeter will consume gas past the limit.
-	//
-	// NOTE: This must exist in a separate defer function for the above recovery
-	// to recover from this one.
+	// Must not panic in this defer function, because tx is already executed and committed,
+	// panic here will return error result for a successfully executed tx.
 	defer func() {
+		gasConsumed := sdkCtx.GasMeter().GasConsumedToLimit()
+		remaining := sdkCtx.BlockGasMeter().GasRemaining()
+		if gasConsumed > remaining {
+			gasConsumed = remaining
+		}
```

Comment:
`gasConsumed` makes this super confusing to read. I'd suggest renaming the variable to `gasToConsume`. (assuming I understand these changes correctly)